File: 0_1.baekjoon/priority-queue/java/BOJ22254.java
import java.util.*;
import java.io.*;

class BOJ22254 {
    static int N, X;
    static int[] gifts;
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        X = Integer.parseInt(st.nextToken());

        gifts = new int[N];

        st = new StringTokenizer(br.readLine());

        for (int i=0;i<N;i++) {
            gifts[i] = Integer.parseInt(st.nextToken());
        }

        int left = 0, right = N + 1;
        int ans = right;
        while (left + 1 < right) {
            int mid = left + (right - left) / 2;

            boolean result = underLimitTimes(mid);
            if (!result) {
                left = mid;
            } else {
                right = mid;
                ans = right;
            }
        }
        System.out.println(ans);

    }
    private static boolean underLimitTimes(int k) {
        PriorityQueue<Integer> pq = new PriorityQueue<>();
        
        int i = 0;
        for (; i < Math.min(N, k); i++){
            pq.add(gifts[i]);
        }

        while (!pq.isEmpty()) {
            int gift = pq.poll();
            
            if (gift > X) return false;

            if (i < N) {
                pq.add(gifts[i++] + gift);
            }
        }

        return true;
    }
}

File: 0_1.baekjoon/priority-queue/python/BOJ22254.py
from sys import stdin
import heapq
readline = stdin.readline

N, X = map(int, readline().split())
gifts = list(map(int, readline().split()))

def solve():
    left, right = 0, N

    while left + 1 < right:
        mid = left + (right - left) // 2

        if get_production_time(mid) > X:
            left = mid
        else:
            right = mid

    print(right)

def get_production_time(limit):
    lines = gifts[:limit]
    heapq.heapify(lines)

    result = 0
    idx = limit

    while lines:
        current = heapq.heappop(lines)

        result = max(result, current)

        if idx < N:
            heapq.heappush(lines, current + gifts[idx])
            idx += 1

    return result

solve()